core/src/main/java/org/apache/calcite/rex/RexAnalyzer.java

```diff
@@ -128,6 +128,10 @@ public RexAnalyzer(RexNode e, RelOptPredicateList predicates) {
     }
 
     @Override public Void visitCall(RexCall call) {
+      if (!RexUtil.isDeterministic(call)) {
```

Comment:
`RexUtil.isDeterministic` walks over the full expression tree which might result in O(N^2) behaviors. Did you mean `call.getOperator().isDeterministic()` here?

Reply:
Oops, you're right.

Reply:
There's a question what to do with `AND(randbool(), randbool(), randbool(), randbool())` :-/

Reply:
Note that VariableCollector is instance of RexVisitorImpl. So it will visit the operand and then return null.